libp2p/protocols/rendezvous.nim

```diff
@@ -0,0 +1,648 @@
+
+proc discover(rdv: RendezVous, conn: Connection, d: Discover) {.async.} =
+  trace "Received Discover", peerId = conn.peerId, ns = d.ns
+  if d.ns.len notin 0..255:
+    await conn.sendDiscoverResponseError(InvalidNamespace)
+    return
+  var limit = min(DiscoverLimit, d.limit.get(DiscoverLimit))
+  let
+    nsSalted = d.ns & rdv.salt
+    cookie =
+      if d.cookie.isSome():
+        try:
+          Cookie.decode(d.cookie.get()).get()
+        except CatchableError:
+          await conn.sendDiscoverResponseError(InvalidCookie)
+          return
+      else: Cookie()
+    indexes =
+      if d.ns != "":
+        try:
+          rdv.indexes[nsSalted]
+        except KeyError:
+          await conn.sendDiscoverResponseError(InvalidNamespace)
+          return
+      else: toSeq(rdv.registered.low()..rdv.registered.high())
+  if indexes.len() == 0:
+    await conn.sendDiscoverResponse(@[], Cookie())
+    return
+  var offset = indexes[^1]
+  let n = Moment.now()
+  let s = collect(newSeq()):
+      for index in indexes:
+        var reg = rdv.registered[index]
+        if reg.expiration < n or index.uint64 <= cookie.offset: continue
+        if limit == 0:
+          offset = index
+          break
+        limit.dec()
+        reg.data.ttl = some((reg.expiration - Moment.now()).seconds.uint64)
+        reg.data
+  await conn.sendDiscoverResponse(s, Cookie(offset: offset.uint64, ns: d.ns))
+
+proc advertisePeer(rdv: RendezVous,
+                   peer: PeerId,
+                   msg: seq[byte],
+                   sema: AsyncSemaphore) {.async.} =
+  proc advertiseWrap() {.async.} =
+    try:
+      let conn = await rdv.switch.dial(peer, RendezVousCodec)
+      defer: await conn.close()
+      await conn.writeLp(msg)
+      let
+        buf = await conn.readLp(4096)
+        msgRecv = Message.decode(buf).get()
+      if msgRecv.msgType != MessageType.RegisterResponse:
+        trace "Unexpected register response", peer, msgType = msgRecv.msgType
+      elif msgRecv.registerResponse.isNone() or
+           msgRecv.registerResponse.get().status != ResponseStatus.Ok:
+        trace "Refuse to register", peer, response = msgRecv.registerResponse
+    except CatchableError as exc:
+      trace "exception in the advertise", error = exc.msg
+    finally:
+      rdv.sema.release()
+      sema.release()
+  discard await advertiseWrap().withTimeout(5.seconds)
+
+proc advertise*(rdv: RendezVous,
+                ns: string,
+                ttl: uint64 = MinimumTTL) {.async.} =
+  let
+    sprBuff = rdv.switch.peerInfo.signedPeerRecord.encode()
+    sema = newAsyncSemaphore(1)
+  if sprBuff.isErr() or ns.len notin 1..255 or ttl notin MinimumTTL..MaximumTTL:
+    return
+  let
+    r = Register(ns: ns, signedPeerRecord: sprBuff.get(), ttl: some(ttl))
+    msg = encode(Message(msgType: MessageType.Register, register: some(r)))
+  rdv.save(ns, rdv.switch.peerInfo.peerId, r)
+  for peer in rdv.peers:
+    await rdv.sema.acquire()
+    sema.forceAcquire()
```

Comment:
Is this a hacky way to wait for all of them to be done? Lol
Instead you could every advertisePeer in a seq[Future[void]] and use await allFutures or similar

Reply:
No, it's a hacky way to send only five requests at the same time. But I agree that the two semaphores acquire could be better in advertisePeer than here.

Reply:
You only need a single semaphore: `rdv.sema.acquire` at the beginning of `advertisePeer`, release it at the end.
Start every `advertisePeer` at once, and wait for all of them here